### PR TITLE
Improve parser automatic paragraph generation

### PR DESCRIPTION
**Pull request by heronils**
An elegant solution for #3854.

* The first commit changes one (!) regex and with this change all unwanted p´s are gone :-) It also adds a CSS rule in order to fix the resulting padding issues in the sidebar.
* The second commit can be ignored, i replace those changes in the third commit.
* The third commit introduces a (hopefully) smart removal of whitespaces after opening and before closing tags of an element. It contains the most code changes. It is needed because of the new way to interpret newlines after an opening tag.

It´s all done in the parser.

Please see the descriptions in the commit comments and in the source code for further infos.

May be possible that we need to tweak some more CSS rules or a few template tiddlers here and there but so far everything seems to work.

For the records: After these changes we are down from 76 to 66 errors in the W3C html validator. I tested a complete wiki document, removing the storage and core functionality divs, and the styles from the head - otherwise the document is too huge for the validator - and i added a html5 doctype.

### CODE REVIEW

**Comment by Jermolene** [comment]
Thanks @heronils, I appreciate the copious comments.

Fixing paragraph generation has two parts: avoiding generating extraneous paragraph tags where they are not needed while also correctly generating paragraph tags where they are needed. As far as I can tell, this PR only addresses the first part.

It's already possible to avoid automatic paragraph generation (albeit the techniques are somewhat clumsy and unintuitive) so it's the second part that is of primary interest. I now believe that it is not possible to do paragraph generation at parse time, I think it has to be done at render time, via a new widget that conditionally generates an HTML P element.

**Comment by Jermolene** [comment]
Hi @heronils apologies I only just caught up on #3854, I'd forgotten we'd already been discussing this issue.

**Comment by heronils** [comment]
Hi, @Jermolene 

> As far as I can tell, this PR only addresses the first part.

Could you give an example where this does not do what you want?

> It's already possible to avoid automatic paragraph generation (albeit the techniques are somewhat clumsy and unintuitive)

If you are referring to \whitespace trim, yes, it is clumsy. It is not very useful when mixing text and elements. We should look that this trimming is done automatically, like in commit 3 of this PR.

> I think it has to be done at render time, via a new widget that conditionally generates an HTML P element.

Yes, you commented about this earlier in two locations, and i also thought this would be the simplest approach. I tried a few things doing it at render time and i failed two times. I now know that it would be necessary to have the widget tree (or at least some parts) created _before_ the dom nodes are created and appended. The reason is that we need to dig into widgets like macrocall and transclusion.

But then i found this PR´s solution and i consider it to be much simpler.

If anyone is interested, i can provide a short code walkthrough explaining what this PR does by pointing to concrete code locations.

File: core/modules/parsers/wikiparser/wikiparser.js [hunk 1]
```diff
@@ -142,6 +142,23 @@ WikiParser.prototype.skipWhitespace = function(options) {
...
+after those newlines. That is because the first code line in the pre may
+intentionally start with an indent. We (have to) assume here that the user does not
+intentionally start his pre´s with newlines. For these cases there is css padding.
+*/
+WikiParser.prototype.skipNewlines = function() {
+	var newlinesRegExp = /((?:[^\S\n]*\r?\n)+)/mg;
+	newlinesRegExp.lastIndex = this.pos;
+	var newlinesMatch = newlinesRegExp.exec(this.source);
+	if(newlinesMatch && newlinesMatch.index === this.pos) {
+		this.pos = newlinesRegExp.lastIndex;
```
[the remark this answers is not in the document]
Comment (pmario) [reply core/modules/parsers/wikiparser/wikiparser.js]: As far as I know, .exec() changes newlinesRegExp.lastIndex;

**Comment by pmario** [comment]
@Jermolene .. 

>Fixing paragraph generation has two parts: avoiding generating extraneous paragraph tags where they are not needed while also correctly generating paragraph tags where they are needed. As far as I can tell, this PR only addresses the first part.

I think that's not true. Please read the commit info: https://github.com/Jermolene/TiddlyWiki5/pull/4195/commits/69089cc47986a0c98092486283714ff7c38ca434

I think this PR solves many problems, we have with "minimum readability" and unwanted p-tag creation. ... 

For extended readability we will need to add a lot of `\trim whitespace` pragmas into our templates.

File: core/modules/parsers/wikiparser/wikiparser.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (heronils) [reply core/modules/parsers/wikiparser/wikiparser.js]: yes, if you use the _g_ flag ('search globally') then _lastIndex_ [will be changed by test() and exec()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/lastIndex).

```
>> var pat = /a/g;  // 'g' is relevant
undefined
>> pat.lastIndex
0
>> pat.exec('aaa');
Array [ "a" ]
>> pat.lastIndex
1
>> pat.exec('aaa');
Array [ "a" ]
>> pat.lastIndex
2
>> pat.lastIndex = 0;
0
>> pat.lastIndex
0
```

**Comment by heronils** [comment]
For the records, it also works with block transclusion (and block macros) when the contents of the transcluded tiddler are wrapped in an element, eg.

transcluded_tiddler:

```
<div>
Hi</div>
```

transcluding tiddler:

```
p

{{transcluded_tiddler}}

p
```

results in

```
<p>p</p><div>Hi</div><p>p</p>
```

(no p around the div)

---

We could also achieve @Arlen22 s PR by introducing a new content type. This would also be very simple. In the wikiparser.js:

1. add `exports["text/template"] = WikiParser;` at the end.

2. replace

```
// Treat it as a paragraph if we didn't find a block rule
return [{type: "element", tag: "p", children: this.parseInlineRun(terminatorRegExp)}];
```

with

```
if (this.type === 'text/template') {
	return this.parseInlineRun(terminatorRegExp);
} else {
	// Treat it as a paragraph if we didn't find a block rule
	return [{type: "element", tag: "p", children: this.parseInlineRun(terminatorRegExp)}];
}
```

Then the interface developer can choose 'text\template' as content type for the tiddler and gets rid of all autogenerated p´s (he has to write them by hand if he still wants them).

That said, i still like my PR as an addition to the normal wikitext, because one may want to have both, auto p generation as well as the possibility to avoid it for selected elements. It is also very intuitive in my opinion, the placement (or absence) of the p´s relates to the amount of linebreaks: No linebreak --> p outside, one linebreak --> no p, two linebreaks --> p inside.

I would like to have both.

---

Regarding `\whitespace trim`. If i am not wrong, then the main purpose of `\whitespace trim` (after this PR) would be

* if the user wants to remove indents of block html elements at the line start, which were added for better readability, and
* he wants to trim inline whitespace between elements and other elements (or text).

I would like if we had explicit options for this, like `\whitespace trim start` and `\whitespace trim between`, and `\whitespace trim` being an alias for `\whitespace trim start between`. But thats a topic for another PR i guess :-) ).

**Comment by Jermolene** [comment]
Thanks everyone. I still haven't been able to invest the time to properly play with the code and understand it fully, but have a few thoughts:

* Our goal should be to produce HTML that validates with the W3C test suite
* We cannot break backwards compatibility, and yet we need to be able to improve things. I favour adding a mechanism for parameterising content types as described in #345 so that we can have content co-existing that was written for different parser configurations
* We need to add a new file to the test edition that enumerates all the cases we wish to address to communicate more clearly the intended parsing behaviour. We'd start with test cases for the current parser that includes the cases we want to change but with the current target results, and then clone those test cases for the new parser, changing the test case target results as we go

**Comment by heronils** [comment]
@Jermolene , Im just [having a look](https://tiddlywiki.com/dev/#TestingMechanism) at these test cases. Will append to this PR if that is ok for you.

**Comment by heronils** [comment]
Ok, there were 9 failures.

The first three commits make the tests succeed again.
Then i add some extra tests which will succeed with or without this PR (commit 4 to 6).
The rest are tests for this PR.
I tried my best to make it understandable. I hope i succeeded :-)

@jermolene 

> HTML5 compatibility
Agree.

> We cannot break backwards compatibility (...)
In my opinion this PR is (now) quite sensible regarding breaking things. It does not break backwards compatibility. The only change (except the whitespace trimming) is that a start tag followed by a line break now results in the element being parsed in block mode (parsing of its children behaves like before). All the user needs to do to get the wrapping p back, is to remove that linebreak. Then everything (again, except the whitespace trimming) behaves like before.

> vocabulary parameters
I like the idea of vocabulary parameters and the current procedere _may_ become such a parameter for the text/vnd.tiddlywiki content type (but to be honest, i think we should just move on) but i my opinon this PR´s behavior should be the default. We dont need to depreace anything, all we need to do is to inform the users that from now on adding exactly one linebreak after an opening tag will behave different and that no linebreak after the tag will result in the old behavior.

**Comment by heronils** [comment]
The above two commits show how to get a desired p back. The tags in sidebar -> more -> tags should not have their wrapping p removed. So, in the second commit i remove the newline after the `<span class="tc-tag-list-item">` in the _$:/core/ui/TagTemplate_ tiddler and the p is back. The first commit was accidental, but shows how to do it without making changes in the transcluded tiddler.

**Comment by heronils** [comment]
Hi again,

I am currently checking what this PR will (or will not) break. Currently i compare the static html of editions\tw5.com\ generated from the main branch with that generated from this branch. This will take one or two weeks (aka, expect three) as these files are 11mb and im also currently in the process of switching apartment. Will post updates.

**Comment by heronils** [comment]
I need one week more, sorry.

**Comment by heronils** [comment]
Closing in favour of  #4290

**Pull request closed without merging**